File: Problems/No.3_Longest_Substring_Without_Repeating_Characters/longest_substring_without_repeating_characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        
        int s_length = s.length();
      
        // to store answer, strting point of unrepeated region, and the characters' last index
        int LSWRC_ans = 0;
        int unrepeated = 0;
        map<char, int> index;
        
        // start compute
        for (int i=0; i<s_length; i++) {
            
            // check if the character appeared before
            if (index.find(s[i]) == index.end()) index[s[i]] = i;
            else {
                
                // calculate the longest substring without repeating characters
                LSWRC_ans = max(LSWRC_ans, i-unrepeated);
              
                // update the starting point of unrepeated region
                if (unrepeated < index[s[i]]+1) unrepeated = index[s[i]]+1;
              
                // update the characters' last index
                index[s[i]] = i;
            
            }
            
        }
        LSWRC_ans = max(LSWRC_ans, s_length-unrepeated);
        
        return LSWRC_ans;
        
    }
};

File: Problems/No.3_Longest_Substring_Without_Repeating_Characters/longest_substring_without_repeating_characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        
        # to store the answer and the unrepeated substring
        LSWRC_ans = 0
        unrepeated = ""
        
        # start compute
        for c in s:
            
            # check if c in unrepeated substring
            if c not in unrepeated:
                unrepeated += c
            else:
              
                # update answer and unrepeated substring
                LSWRC_ans = max(LSWRC_ans, len(unrepeated))
                unrepeated = unrepeated[unrepeated.index(c)+1:] + c
     
        return LSWRC_ans if LSWRC_ans > len(unrepeated) else len(unrepeated)
